File: lab3/src/main/java/org/oop/lab3/entities/Player.java
package org.oop.lab3.entities;

import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import lombok.Data;

@JsonPropertyOrder({
    "id",
    "name",
    "height",
    "jersey_number",
    "team",
    "type",
    "point_guard",
    "center"
})
@Data
@Entity
@Table(
    name="players",
    uniqueConstraints = @UniqueConstraint(columnNames = {"name"})
)
public class Player implements Playable{

    public Player() {}

    public Player(String name, int height, int jersey_number, Type type) {
        this.name = name;
        this.height = height;
        this.jerseyNumber = jersey_number;
        this.type = type;
    }

    public enum Type {
        pointGuard,
        center
    }

    @Id
    @GeneratedValue(strategy=GenerationType.IDENTITY)
    @Column(name="id")
    private Long id;

    @OneToOne(mappedBy="player", cascade=CascadeType.ALL, optional=true, orphanRemoval=true)
    @JsonManagedReference
    @Valid
    private Center center;

    @OneToOne(mappedBy="player", cascade=CascadeType.ALL, optional=true, orphanRemoval=true)
    @JsonManagedReference
    @Valid
    private PointGuard pointGuard;

    @Column(name="name")
    @NotNull(message="Имя не должно быть пустым")
    @NotBlank(message="Имя не должно быть пустым")
    @NotEmpty(message="Имя не должно быть пустым")
    private String name;

    @Column(name="height")
    @Min(value = 150, message="Рост должен быть хотя бы 150")
    @Max(value = 230, message="Рост не должен быть больше 230")
    private Integer height;

    @Column(name="jersey_number")
    @Min(value = 0, message="Номер должен быть хотя бы 0")
    @Max(value = 99, message="Номер не должен быть больше 99")
    private Integer jerseyNumber;

    @Column(name="type")
    @Enumerated(EnumType.ORDINAL)
    private Type type;

    @OnDelete(action=OnDeleteAction.SET_NULL)
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "team_id")
    private Team team;

    public String play() {
        if (center != null) {
            return center.play();
        } else {
            return pointGuard.play();
        }
    }

    public String train() {
        if (center != null) {
            return center.train();
        } else {
            return pointGuard.train();
        }
    }

    public String printInfo() {
        if (center != null) {
            return String.format("Name: %s\nHeight: %d\nJersey number: %d\n", name, height, jerseyNumber) + center.printInfo();
        } else {
            return String.format("Name: %s\nHeight: %d\nJersey number: %d\n", name, height, jerseyNumber) + pointGuard.printInfo();
        }
    }
}

File: lab4/backend/database/models/Player.py
from typing import List, TYPE_CHECKING
from typing import Optional
from sqlalchemy import ForeignKey
from sqlalchemy import String
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column 
from sqlalchemy.orm import relationship
from .team import Team
from ..base import Base
from enum import Enum

if TYPE_CHECKING:
    from .center import Center
    from .point_guard import PointGuard

class Type(int, Enum):
    point_guard = 0
    center = 1
class Player(Base):
    __tablename__ = "players"
    
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(30), unique=True)
    height: Mapped[int]
    jersey_number: Mapped[int]
    type: Mapped[int]
    games_played: Mapped[int] = mapped_column(nullable=True)
    team_id: Mapped[int] =  mapped_column(ForeignKey("teams.id", ondelete="SET NULL"), nullable=True)

    center: Mapped[Optional["Center"]] = relationship(back_populates="player", cascade="all, delete")
    point_guard: Mapped[Optional["PointGuard"]] = relationship(back_populates="player", cascade="all, delete")
    
    def play(self):
        return self.getPointGuard().play() if self.type == Type.point_guard else self.getCenter().play()
    
    def train(self):
        return self.getPointGuard().train() if self.type == Type.point_guard else self.getCenter().train()
    
    def getBasicInfo(self):
        return f"Name: {self.name}\nHeight:: {self.height}\nJersey number: {self.jersey_number}\nGames player: {self.games_played}"
    
    def printInfo(self):
        return self.getPointGuard().printInfo() if self.type == Type.point_guard else self.getCenter().printInfo()

    def __repr__(self) -> str:
        return f"""Player(id={self.id!r},
            name={self.name!r},
            height={self.height!r},
            jersey_number={self.jersey_number!r},
            type={self.type!r},
            team_id={self.team_id!r},
            center={self.center!r},
            point_guard={self.point_guard!r})
        """
        
    def getName(self):
        return self.name
    
    def getCenter(self):
        return self.center
    
    def getPointGuard(self):
        return self.point_guard
    